incorporate processing of Google Form tweet entries

jrgant opened the pull request:
@malcolmbarrett We should probably chat before merging this, but I've come up with a solution for processing tweets from the Google Form. Required some changes to the work flow, but I've tested it the best I can with the sandbox account.

jrgant committed:
incorporate processing of Google Form tweet entries

jrgant commented:
Here's a general description of the new workflow in `action_auto_tweet()` and `post_tweet_queue()`

1. Pull the Google Sheet attached to the Google Form for tweet entry (tweet_entry_responses)
2. Pull a second Google Sheet that contains the IDs for tweets that have already been posted (tweet_history)
3. Subset tweet_entry_responses to IDs that don't appear in tweet_history, and select one at random.
4. Tweet.
5. Write the tweet's ID to history and update the tweet_history Drive file.
6. ...
7. Profit.

A couple additional notes: 

- TweetIDs are simply the collapsed timestamps from tweet_entry_responses. Technically, not 100% foolproof re: duplication, but for all practical purposes, I doubt two people will ever posts at exactly the same second.

- We no longer re-upload the tweet queue (ie, tweet_form_responses) because replacing it via `drive_update()` unlinks it from the Google Form.

malcolmbarrett commented:
Thanks, @jrgant! Could you give us a demo using the sandbox account at this week's meeting? I will review the code after I see it in action

jrgant commented:
Sure thing! Sounds good.

jrgant committed:
correct tweet history filenames

jrgant committed:
if all tweets tweeted, restart tweet history

malcolmbarrett commented:
@jrgant ping me when you are done with your updates, and I will find a time to review it!

jrgant commented:
@malcolmbarrett Will do. Need to finish testing. I think I ran into Twitter's rate limit for posting while I was making additional changes and haven't had a chance to try again.

jrgant committed:
minor changes to object and function naming

jrgant committed:
remove redundant code

jrgant committed:
add flag for testing, force day to be weekday

jrgant commented:
@malcolmbarrett I just ran a successful test for restarting the tweet history when all tweets in the library had already been posted. Still not sure what the issue was last time, but stepping through the processing, looks like things should be okay. 

Added some additional code today to force `todays_date` to read as a weekday so we can manipulate for testing purposes.

malcolmbarrett commented on inst/scripts/test_auto_tweet.R:
```diff
@@ -20,7 +20,7 @@ is_tweet_length <- function(.x, n = 280) {
-safe_action_auto_tweet <- action_safely(action_auto_tweet)
+# if testing post_tweet_of_type manually on a weekend, set apply_test_dat = TRUE
+# in action_auto_tweet() below
+safe_action_auto_tweet <- action_safely(action_auto_tweet(twitter_token = sandbox_token()))
```
Is this right? Doesn't `action_safley()` take a function and return a function? Are you sure the `twitter_token` argument shouldn't be in the line below, e.g 
```r
safe_action_auto_tweet <- action_safely(action_auto_tweet)
safe_action_auto_tweet(twitter_token = sandbox_token())
```

jrgant replied:
I've made this update to the code, but I ended up not being able to run a test because I don't have the SER email account's secret. If I remember, I think that might be why I did it the original way.

Can you share that info with me?

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -1,16 +1,29 @@
 
-  tweet_data <- tweet_data %>%
-    dplyr::mutate(tweeted = ifelse(text == tweet, TRUE, tweeted))
+  tweet_datac <- tweet_data %>%
```
Let's just call this `tweet_data` rather than `tweet_datac`. It's happening in the function namespace so there's no need to worry about overwriting the original data

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -59,19 +72,36 @@ blackout_tweet <- function(tweet_data = tweet_queue) {
+    googledrive::as_id()
+  googledrive::drive_download(tweet_hist_id, type = "csv", overwrite = TRUE)
+  tweet_hist <- readr::read_csv("ser_tweet_history.csv")
+  tweet_hist_list <- as.character(tweet_hist$tweet_id)
```
Specify that `tweet_id` is a character in `readr::read_csv()` instead of manually converting

malcolmbarrett replied:
Also, I'd prefer to call this something like `tweet_hist_ids` because it is a vector, not a list

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -80,18 +110,27 @@ action_auto_tweet <- function(twitter_token = ser_token, google_drive_auth = dri
+    todays_date <- lubridate::wday(Sys.Date(), label = TRUE) %>%
+      as.character()
+  }
+
```
Good idea, but let's do this instead:

1) create and export a helper function, something like `get_day()` that returns the lubridate code here. 
2) change the function argument `apply_test_date` to `todays_date ` with a default value of `get_day()`. 
3) then we can delete this section. 

This simplifies the code a bit and then in the tests you can write "Mon" for `todays_date`

malcolmbarrett replied:
actually, I don't think we'd need to export `get_day()`

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -80,18 +110,27 @@ action_auto_tweet <- function(twitter_token = ser_token, google_drive_auth = dri
-  if (retweet_day) post_tweet_of_type <- post_retweet
+  if (retweet_day) {
+    post_tweet_of_type <- post_retweet(twitter_token = twitter_token)
+    }
```
Please put this on a single line (e.g. `if (retweet_day) post_tweet_of_type <- post_retweet(twitter_token = twitter_token)`

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -80,18 +110,27 @@ action_auto_tweet <- function(twitter_token = ser_token, google_drive_auth = dri
-    post_tweet_of_type <- post_tweet_queue
+    post_tweet_of_type <- post_tweet_library(tweet_data = tweet_library,
+                                             past_tweets = tweet_hist_list,
+                                             twitter_token = twitter_token)
```
Could you re-style this using the tidyverse format (https://style.tidyverse.org/syntax.html#long-lines)?

e.g.

```r
    post_tweet_of_type <- post_tweet_library(
      tweet_data = tweet_library,
      past_tweets = tweet_hist_list,
      twitter_token = twitter_token
    )
```

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -107,17 +146,23 @@ action_auto_tweet <- function(twitter_token = ser_token, google_drive_auth = dri
   } else {
-    tweet_queue <- post_tweet_of_type(tweet_queue)
+    tweet_library_status <- post_tweet_of_type
+  }
```
It looks like you changed the code such that `post_tweet_of_type` is not a function but the result of a post. The first part of this logic treats it as a function still, and the second part creates an object you need later, `tweet_library_status`. If it's a retweet day, this will fail, right? Even if you change the first error, `post_retweet()` doesn't return anything called `just_tweeted`, so it will fail later.

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -80,18 +110,27 @@ action_auto_tweet <- function(twitter_token = ser_token, google_drive_auth = dri
+                                             past_tweets = tweet_hist_list,
+                                             twitter_token = twitter_token)
   } else {
     post_tweet_of_type <- blackout_tweet
```
If `post_tweet_of_type` is no longer a function, you need to actually run `blackout_tweet`, e.g.

`post_tweet_of_type <- blackout_tweet()`

malcolmbarrett replied:
This also needs to happen on line 143

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -1,16 +1,29 @@
+  rtweet::post_tweet(tweet_datac$Tweet, token = twitter_token)
+
+  invisible(list(restart_history = restart_history,
+                 just_tweeted = tweet_datac$id_transform))
```
Could you restyle this? E.g.

```r
  invisible(
    list(
      restart_history = restart_history,
      just_tweeted = tweet_datac$id_transform
    )
  )
```

malcolmbarrett commented on R/auto_tweet.R:
```diff
@@ -107,17 +146,23 @@ action_auto_tweet <- function(twitter_token = ser_token, google_drive_auth = dri
+    tweet_hist <- data.frame(tweet_id = tweet_library_status$just_tweeted)
+  } else {
+    tweet_hist <- data.frame(tweet_id = c(tweet_hist_list,
+                                          tweet_library_status$just_tweeted))
```
let's put the second data.frame all on one line, e.g.

`tweet_hist <- data.frame(tweet_id = c(tweet_hist_list, tweet_library_status$just_tweeted))`

jrgant replied:
Putting this on one line turned it into a long line. I decided to do the following:

```
tweet_hist <- data.frame(
   tweet_id = c(tweet_hist_list, tweet_library_status$just_tweeted)
)
```
Let me know if you still prefer it on one-line.

malcolmbarrett reviewed (changes requested):
Thanks!! This is a lot of styling stuff, but I suspect the rate issue you had may have had to do with the changes you made to `post_tweet_of_type`. This was originally a function, which was processed and called once. Because you changed some of the arguments to what is now `post_tweet_library()`, that might not be possible anymore. Make sure each instance of `post_tweet_of_type` is the result of function and not a `function` object. Also, make sure it returns anything you need later in the script, e.g. you call `tweet_library_status$just_tweeted` but only `post_tweet_library()` returns `just_tweeted`, so that will fail if `blackout_tweet()` or `post_retweet()` are called.

Also, since `post_tweet_of_type` no longer a post function, I think we should change the name to `tweet_library_status` right away. Then you can remove the logic on lines 145-150.

jrgant committed:
rename tweet_datac to tweet_data

jrgant committed:
rename tweet_hist_list to tweet_hist_ids and specify character variable type in read_csv

jrgant committed:
put if statement on single line

jrgant committed:
fix long lines

jrgant committed:
address style request

jrgant committed:
add get_day() helper function and update date checking

jrgant committed:
restyle code

jrgant committed:
move twitter_token to safe_action_auto_tweet

malcolmbarrett commented:
This is looking good but the main thing I see is that `post_tweet_of_type` is no longer a function, and so you'll need to address that. A simple way would be to continue to treat it as a function by having `post_tweet_library()` return a function that then posts a tweet. If `post_tweet_of_type` is not going to be a function we need to 1) change the name to something more reflective of the object that's returned 2) adjust the workflow so that blackouts and retweets are being called instead of `post_tweet_library()` where relevant and 3) make sure the code is cleaned up to reflect this all, e.g. no calls to `post_tweet_of_type()` since it's not a function. 

It seems simpler to me to instead continue to have it be a function but it's up to you

malcolmbarrett commented:
I'll resolve the conflict when we're ready to merge. I slightly changed `action_safley()` to include the name of the script that is throwing an error

jrgant committed:
update man for action_auto_tweet

jrgant committed:
add sandbox_token()

jrgant committed:
change post_tweet_of_type back to a function

jrgant commented:
Alright, pushed a few commits, most notably one to revert `post_tweet_of_type` to a function, as you suggested.

I also added a `sandbox_token()` function to the package. However, I'm still getting an error when I try to test: "Error: Must supply either `key` and `secret` or `path`". Any ideas? As far as I can tell, the tokens/keys are visible in my local environment.

jrgant committed:
fix post_tweet_of_type

jrgant committed:
fix/update auto_tweet inputs and emails

jrgant commented:
Alright, I got everything tested, and the script seems to be working.

The only weirdness now is I'm getting error emails that read: `Error in .f(...): could not find function ".f"`. However, the tweets still go through.

jrgant committed:
resolve conflicts

jrgant committed:
re-resolve conflicts

malcolmbarrett commented:
Will merging this break the auto tweet since the data here is limited to two fake tweets?

malcolmbarrett commented:
Also, are you still getting error emails about `.f`?

malcolmbarrett committed:
use tibbles

malcolmbarrett committed:
use sandbox in test

malcolmbarrett committed:
run styler

malcolmbarrett committed:
Merge branch 'master' into master

jrgant commented:
> Will merging this break the auto tweet since the data here is limited to two fake tweets?

I don't think it should, but can you clarify the concern here? I can check for the `.f` errors tomorrow.

malcolmbarrett commented:
`"ser_tweet_library.csv"` only contains two fake tweets. If I merge this right now, then `action_auto_tweet()` uses that file, right? That would surely result in a couple of nonsense tweets on the SER account

jrgant commented:
I think what we need to do is change the name of the `tweet_queue` Google Sheet to `ser_tweet_library`, right?

Looks to me that `action_auto_tweet` will look for the sheet in Drive, download it, and save it to `ser_tweet_library.csv`, overwriting the existing CSV of that name.

I thought that nonsense CSV was only on my local, though. Are you seeing it elsewhere?

malcolmbarrett commented:
I'm confused. I thought this was supposed to point to a sheet generated by Google Forms.

jrgant commented:
There was supposed to be a standardized Google Form that added entries to the ser_tweet_library sheet, and the changes here were supposed to account for that process (and to update language to refer to the tweet library instead of the tweet queue). However, I don't see that form in the SER Google Drive anymore. I'll bring this up in today's Analytics team meeting to see if anyone remembers who owns that form. If need be, I'll revisit these updates and get back to you.

malcolmbarrett commented:
\style

malcolmbarrett commented:
/style

jrgant committed:
Merge branch 'SERTwitter:master' into master

jrgant committed:
build in check for tweet recurrence flag

The Library Team requested that we be able to have some tweets allowed
to recur. Therefore, we added a column called "Recurrence" to the Tweet
Entry form, which takes "Yes" if the tweet is allowed to recur and "No"
if not. The new check occurs within `post_tweet_library()` and removes
tweets with `Recurrence == "Yes"` from the list IDs to be omitted due to
having been posted once already.

jrgant commented:
I think we worked the remaining kinks out in October 2021 on that call. I just added some minor functionality requested by the Library team in commit `fe5681d`. Basically, it checks for whether the tweet is allowed to recur or not, and if so, does not omit the tweet's ID just because it shows up in tweet_history.

malcolmbarrett commented:
I'm on paternity leave so I'll let you make the final call on if it's ready to merge. Don't break it lol. Let me know if you don't have merge privileges

jrgant commented:
Sounds good, thanks. Enjoy your leave!

On Fri, Mar 11, 2022 at 11:46 AM Malcolm Barrett ***@***.***>
wrote:

> I'm on paternity leave so I'll let you make the final call on if it's
> ready to merge. Don't break it lol. Let me know if you don't have merge
> privileges
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SERTwitter/ser/pull/10#issuecomment-1065292747>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIRB2KDPQU3KCK4XLN3DIDU7N2HFANCNFSM4KO7XXXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
